torch/optim/lars.py

```diff
@@ -0,0 +1,214 @@
+
+LARS.__doc__ = r"""Implements LARS algorithm.
+
+    For further details regarding the algorithm we refer to `Large Batch Training of Convolutional Networks`_.
+    """ + r"""
+    Args:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float): learning rate
+        momentum (float, optional): momentum factor (default: 0)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        nesterov (bool, optional): enables Nesterov momentum (default: False)
+        trust_coefficient (float, optional): coefficient for computing LR (default: 0.001)
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-8)
+        {maximize}
+        {differentiable}
+
+    .. _Large Batch Training of Convolutional Networks:
+        https://arxiv.org/abs/1708.03888
+
+    """.format(maximize=_maximize_doc, differentiable=_differentiable_doc) + r"""
+
+    Example:
+        >>> # xdoctest: +SKIP
+        >>> optimizer = torch.optim.LARS(model.parameters(), lr=0.1, momentum=0.9)
+        >>> optimizer.zero_grad()
+        >>> loss_fn(model(input), target).backward()
+        >>> optimizer.step()
+    """
+
+def lars(
+    params: List[Tensor],
+    grads: List[Tensor],
+    momentum_buffer_list: List[Optional[Tensor]],
+    *,
+    lr: float,
+    momentum: float,
+    dampening: float,
+    weight_decay: float,
+    nesterov: bool,
+    trust_coefficient: float,
+    eps: float,
+    maximize: bool,
+):
+    r"""Functional API that performs LARS algorithm computation.
+    See :class:`~torch.optim.LARS` for details.
+    """
+    if torch.jit.is_scripting():
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if not torch.jit.is_scripting():
+        func = _single_tensor_lars
+
+    func(
+        params,
+        grads,
+        momentum_buffer_list,
+        lr=lr,
+        momentum=momentum,
+        dampening=dampening,
+        weight_decay=weight_decay,
+        nesterov=nesterov,
+        trust_coefficient=trust_coefficient,
+        eps=eps,
+        maximize=maximize,
+    )
+
+
+def _single_tensor_lars(
+    params: List[Tensor],
+    grads: List[Tensor],
+    momentum_buffer_list: List[Optional[Tensor]],
+    *,
+    lr: float,
+    momentum: float,
+    dampening: float,
+    weight_decay: float,
+    nesterov: bool,
+    trust_coefficient: float,
+    eps: float,
+    maximize: bool,
+):
+    for i, param in enumerate(params):
+        d_p = grads[i] if not maximize else -grads[i]
+
+        p_norm = torch.norm(param.data)
+        g_norm = torch.norm(d_p.data)
+
+        if weight_decay != 0:
+            # LARS scaling:
+            if p_norm * g_norm > 0:
+                lars_lr = trust_coefficient * p_norm / (g_norm + p_norm * weight_decay + eps)
+
+                d_p = d_p.add(param, alpha=weight_decay)
+                d_p.mul_(lars_lr)
```

Comment:
```suggestion
                d_p = d_p * lars_lr
```
avoid inplace updates

Reply:
Ok. There're some other in-place updates. 

```suggestion
                param.add_(d_p, alpha=-lr)
```